File: DemoClient/projects/ng-realtime-database/src/lib/models/prefilter/order-by-prefilter.ts
import {IPrefilter} from './iprefilter';
import {ArrayHelper} from '../../helper/array-helper';

export class OrderByPrefilter<T> implements IPrefilter<T> {
  prefilterType = 'OrderByPrefilter';
  selectFunction: (x: T) => any;
  selectFunctionString: string;
  descending: boolean;
  contextData: { [key: string]: string };

  constructor(selectFunction: (x: T) => any, descending: boolean = false, contextData?: { [key: string]: any }) {
    this.selectFunction = selectFunction;
    this.selectFunctionString = selectFunction.toString();
    this.descending = descending;

    if (contextData) {
      this.contextData = {};

      for (const key of Object.keys(contextData)) {
        this.contextData[key] = JSON.stringify(contextData[key]);
      }
    }
  }

  public execute(values: T[]) {
    return ArrayHelper.orderBy(values, this.selectFunction, this.descending);
  }

  public hash() {
    return `${this.prefilterType},${this.selectFunctionString},${this.descending},${JSON.stringify(this.contextData)}`;
  }
}

File: RealtimeDatabase/Internal/Prefilter/OrderByPrefilter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using JavaScriptEngineSwitcher.Core;
using RealtimeDatabase.Helper;

// ReSharper disable PossibleMultipleEnumeration

namespace RealtimeDatabase.Internal.Prefilter
{
    public class OrderByPrefilter : IPrefilter
    {
        public OrderByPrefilter()
        {
            engine = JsEngineSwitcher.Current.CreateDefaultEngine();
        }

        private readonly IJsEngine engine;

        public string SelectFunctionString { get; set; }

        public object[] ContextData { get; set; }

        public bool Descending { get; set; }

        private Func<object, IComparable> keySelector;

        public IEnumerable<object> Execute(IEnumerable<object> array)
        {
            if (array.Any())
            {
                if (keySelector == null)
                {
                    keySelector = SelectFunctionString.CreatePredicateFunction(ContextData, engine);
                }

                return Descending ? array.OrderByDescending(keySelector) : array.OrderBy(keySelector);
            }

            return array;
        }

        public void Dispose()
        {
            engine.Dispose();
        }
    }
}
